QDEIM.C

```diff
@@ -15,46 +15,418 @@
 #include "mpi.h"
 #include <cmath>
 
+#include <vector>
+#include <set>
+#include <algorithm>
+
+/* Use Autotools-detected Fortran name-mangling scheme */
+#define dgesdd CAROM_FC_GLOBAL(dgesdd, DGESDD)
+
+extern "C" {
+void dgesdd(char*, int*, int*, double*, int*,
+             double*, double*, int*, double*, int*,
+             double*, int*, int*, int*);
+}
+
 namespace CAROM {
 
   /* Implement QDEIM algorithm from Zlatko Drmac, Serkan Gugercin, "A
      new selection operator for the discrete empirical interpolation
      method -- Improved a priori error bound and extensions", SIAM
      Journal on Scientific Computing, Volume 38, Issue 2, pages
      A631-A648, 2016.
+     Oversampling uses GappyPOD+E from Peherstorfer, Drmac, Gugercin,
+     "Stability of discrete empirical interpolation and gappy proper
+     orthogonal decomposition with randomized and deterministic
+     sampling points", preprint May 20, 2020.
   */
 void
 QDEIM(const Matrix* f_basis,
       int num_f_basis_vectors_used,
       int* f_sampled_row,
-      int* f_sampled_row_owner,
-      Matrix& f_basis_sampled,
-      int myid)
+      int* f_sampled_rows_per_proc,
+      Matrix& f_basis_sampled_inv,
+      const int myid,
+      const int num_procs,
+      const int num_samples_req)
 {
   // This algorithm determines the rows of f that should be sampled, the
-  // processor that owns each sampled row, and fills f_basis_sampled with the
+  // processor that owns each sampled row, and fills f_basis_sampled_inv with the
   // sampled rows of the basis of the RHS.
 
-  // Compute the number of basis vectors used; this number can't
-  // exceed the number of columns of the matrix.
-  int num_basis_vectors_used = std::min(num_f_basis_vectors_used,
-					f_basis->numColumns());
+  CAROM_VERIFY(num_f_basis_vectors_used == f_basis->numColumns());  // The QR implementation uses the entire matrix.
+  CAROM_VERIFY(f_basis->numColumns() <= num_samples_req && num_samples_req <= f_basis->numRows());
```

Comment:
> I have a similar line in GNAT that is failing due to num_samples possibly being greater than f_basis->numRows for parallel runs since numRows() only returns the number of rows on the current processor. I've created a numDistributedRows function in the tests/gnatTest PR. Perhaps once we merge that PR, we can use that function here too.

@kevinhkhuynh Just to make sure, numDistributedRows return global number of rows, right?

Reply:
@dylan-copeland I can make the change to this line in my test_qdeim branch once test_gnat is finished next week, so this PR isn't held up. How does that sound?

Reply:
That sounds good.